src/sentry/spans/buffer/redis.py

```diff
@@ -3,34 +3,96 @@
 from django.conf import settings
 from sentry_redis_tools.clients import RedisCluster, StrictRedis
 
-from sentry.utils import redis
+from sentry.utils import json, redis
 
 SEGMENT_TTL = 5 * 60  # 5 min TTL in seconds
+TWO_MINUTES = 2 * 60  # 2 min delay in seconds
 
 
 def get_redis_client() -> RedisCluster | StrictRedis:
-    return redis.redis_clusters.get(settings.SENTRY_SPAN_BUFFER_CLUSTER)
+    return redis.redis_clusters.get(settings.SENTRY_SPAN_BUFFER_CLUSTER, decode_responses=False)
 
 
 def get_segment_key(project_id: str | int, segment_id: str) -> str:
     return f"segment:{segment_id}:{project_id}:process-segment"
 
 
+def get_last_processed_timestamp_key(partition_index: int) -> str:
+    return f"performance-issues:last-processed-timestamp:partition:{partition_index}"
+
+
+def get_unprocessed_segments_key(partition_index: int) -> str:
+    return f"performance-issues:unprocessed-segments:partition:{partition_index}"
+
+
 class RedisSpansBuffer:
     def __init__(self):
         self.client: RedisCluster | StrictRedis = get_redis_client()
 
-    def read_segment(self, project_id: str | int, segment_id: str) -> list[str | bytes]:
-        key = get_segment_key(project_id, segment_id)
+    def write_span_and_check_processing(
+        self,
+        project_id: str | int,
+        segment_id: str,
+        timestamp: int,
+        partition: int,
+        span: bytes,
+    ) -> bool:
+        segment_key = get_segment_key(project_id, segment_id)
+        timestamp_key = get_last_processed_timestamp_key(partition)
 
-        return self.client.lrange(key, 0, -1) or []
+        with self.client.pipeline() as p:
+            p.rpush(segment_key, span)
+            p.get(timestamp_key)
```

Comment:
fetching the timestamp to see if the previous value is less than the value we just set on it, which means at least a second has passed and we know we have to process segments. I didn't end up using the `get` param on `set` because it's only available in newer versions of redis.